packages/graph/hash_graph/lib/graph/src/store/postgres/query/table.rs

```diff
@@ -34,6 +43,296 @@ impl TableName {
     }
 }
 
+impl Transpile for Table {
+    fn transpile(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, r#""{}""#, self.as_str())
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum JsonField<'p> {
+    Json(&'p Cow<'p, str>),
+    Text(&'p Cow<'p, str>),
+    Parameter(usize),
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum TypeIds {
+    VersionId,
+    BaseUri,
+    Version,
+    LatestVersion,
+}
+
+impl Transpile for TypeIds {
+    fn transpile(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let column = match self {
+            Self::VersionId => "version_id",
+            Self::BaseUri => "base_uri",
+            Self::Version => "version",
+            Self::LatestVersion => "latest_version",
+        };
+        write!(fmt, r#"."{column}""#)
+    }
+}
+
+macro_rules! impl_ontology_column {
+    ($($name:ident),* $(,)?) => {
+        $(
+            #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+            pub enum $name {
+                VersionId,
+                OwnedById,
+                CreatedById,
+                UpdatedById,
+                Schema(Option<JsonField<'static>>),
+            }
+
+            impl $name {
+                pub const fn nullable(self) -> bool {
+                    match self {
+                        Self::VersionId
+                        | Self::OwnedById
+                        | Self::CreatedById
+                        | Self::UpdatedById => false,
+                        Self::Schema(_) => true,
```

Comment:
Similar to https://github.com/hashintel/hash/pull/1408#discussion_r1025034672
It would be possible to distinguish between the different fields of the ontology types, but for simplicity, we assume that all fields are nullable. Happy to change this, but this will get pretty boiler-platy.

Reply:
I think it's fine for now. It's fine to assume nullability for all property/schema fields. May mean we use another type of join, but really shouldn't be an issue currently